src/components/side-panel/layer-panel/add-layer-button.js

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import React, {useState} from 'react';
+import styled from 'styled-components';
+import {FormattedMessage} from 'localization';
+
+import Tippy from '@tippyjs/react';
+import DropdownList from 'components/common/item-selector/dropdown-list';
+import {Add} from 'components/common/icons';
+import {Button} from 'components/common/styled-components';
+
+const DropdownContainer = styled.div.attrs({
+  className: 'add-layer-menu-dropdown'
+})`
+  .list-selector {
+    border-top: 0;
+    width: max-content;
+    padding: 8px 0;
+    /* disable scrolling, currently set to 280px internally */
+    max-height: unset;
+  }
+
+  .list__header {
+    padding: 8px 20px;
+  }
+
+  .list__item {
+    padding: 8px 20px;
+
+    &:hover {
+      background: ${props => `${props.theme.PURPLE2}22`};
+    }
+    &:hover > div {
+      color: ${props => props.theme.PURPLE2};
+    }
+  }
+
+  .list__item > div {
+    display: flex;
+    align-items: center;
+    flex-direction: row;
+    justify-content: flex-start;
+    line-height: 18px;
+    padding: 0;
+
+    svg {
+      margin-right: 10px;
+    }
+  }
+`;
+
+const AddLayerMenu = styled.div.attrs({
+  className: 'add-layer-menu'
+})`
+  display: flex;
+  flex-direction: column;
+  width: 100%;
+  position: absolute;
+  top: 100%;
+  left: -53px;
+  z-index: 5;
+`;
+
+function AddLayerButtonFactory() {
+  const AddLayerButton = props => {
+    const {datasets, onOptionSelected} = props;
+    const [showAddLayerDropdown, setShowAddLayerDropdown] = useState(false);
+
+    const toggleAddLayerDropdown = () => {
+      setShowAddLayerDropdown(!showAddLayerDropdown);
+    };
+
```

Comment:
import {useCallback} from 'react';

const toggleAddLayerDropdown =  useCallback(() => {
      setShowAddLayerDropdown(!showAddLayerDropdown);
}, [setShowAddLayerDropdown, showAddLayerDropdown]);

Reply:
This doesn't make sense since useCallback is used for things that are passed from parent and showAddLayerDropdown is part of the state of AddLayerButton.

Reply:
In general is always a good practice to use memoization (useCallback, useMemo).

We should try to avoid creating inline objects (style, ...etc) and functions (onClick, ...etc). To improve performance we use  `useMemo` and `useCallback` which can take any dependencies (not just props)

A good read is this: https://itnext.io/6-tips-for-better-react-performance-4329d12c126b

I created an example to show you how to take advantage of `useCallback` appropriately.
Example: https://codesandbox.io/s/interesting-cherry-szu70?file=/src/App.js , if you toggle line 12 with 13 will notice the difference. 
When the `onToggle` is created with `useCallback` the `Button` component will not have unnecessary re-renders. I also use `Button = React.memo(...)` to simulate your specific use case with styled-components

Reply:
Wow, haven't looked at it from this perspective, so when a child just calls the function passed from the parent (and the parent function can change depending on props/state) we should use useCallback.
I'm used to using useCallback inside a child where the function depends on the props.
Thanks for this, and also I think this is also a good read: https://blog.logrocket.com/rethinking-hooks-memoization/ where sprinkling useMemo everywhere might affect performance since useMemo needs to do some checks always and useMemo is run on each render to do the checks.